CorsixTH/CorsixTH
# [WIP] Organise C++ Files and add CPP Unit Testing

Background

Following #1308 I started looking into refactoring the Lua scripts. However CorisxTH would run out of memory and be killed if the scripts were incorrect.

Digging deeper I discovered that we leak around 200MB when running correctly and if we error out can rapidly run into 30GB+ of leaked memory (using Valgrind). This is because we use raw pointers through the program and only return without releasing in many cases when there is an error.

As a roadmap I want to migrate the project from C style pointers to C++ RAII containers (such as unique_ptr). However this requires new unit tests to ensure we don't regress the game as part of the modernisation.

-----
Sorry this is a big PR from nowhere. I wasn't planning on completing this work but things gradually came together over a day.

This PR:
- Adds the Google Test framework as an external project to run the unit tests
- Separates CPP header and source files into respective folders (`inc` and `src`)
- Bumps the minimum C++ standard to C++-14 (allowing us to use `make_unique`). The minimum compiler required now is GCC 4.9+ / MSVC 2013+. However these are both old enough to feel confident about proposing this change.
- Enables target folders on IDEs that support it (such as MSVC) so we can move Unit Test related targets into their own folder.
- Enables using `vcpkg` by default on MSVC. (This allows use of cmake-gui for fresh projects and users who are comfortable providing their own libraries instead should be able to use `-DUSE_VCPKG_DEPS=OFF`)
- Adds post build steps to copy the required folders/file: Bitmaps, Campaigns, Levels, Lua and Corsixth.lua into the build directory using CMake.
- Moves `ConvertLegacySprite` into `sprite.h/cpp` and renames / clarifies the code
- Adds a two new unit tests, `example` which provides a testing framework sanity check. And `SampleTest.h` which was written and checked before the refactoring above took place.
- Adds the Unit Testing step to the Travis Build. I wasn't sure how to do something similar with AppVeyor so left it as is.

As always I'm happy for any / all feedback, suggestions or improvement ideas.

----

Testing:
- Build all targets (`CorsixTH` and `UnitTests`)
- Run the unit test target:
1. `cd CorsixTH/CppTest`
2. `./UnitTests`
- Check CorsixTH still loads correctly. However there have been no changes to the main codebase.

-----

Future work ideas:

One of the most important pieces of work is to resolve the memory leaks throughout the code base by using modern containers. There are three approaches I could take to modernising the codebase in personal preference:

- Adding a new data structure which internally holds the data as a `vector<uint8_t>` which has methods to get a reference or copy of the data as a vector or C array (depending if the caller is working with Lua or not). 
- Converting `uint8_t *` to `std::vector<uint8_t>` within the codebase. Then asking for the vector raw pointer or copying the vector into a new pointer at the last moment when we need to use Lua (depending on whether the Lua function takes ownership). 
- Wrapping all raw pointer arrays in `std::unique_ptr`. This loses cache locality and the benefits of being on the stack. The trade of is we don't need to do the conversion when going to Lua. However if the Lua function does not take ownership we must think about data structures to make sure the `unique_ptr` is not released after we forward to the Lua function.

Alternatively I am happy to consider any other suggestions or ideas?

If I do this work it will be gradual as I can consistently spare only a few hours per week for the upcoming months. I plan on creating two helper methods to convert from raw pointers -> new container (or vice versa) so the codebase can gradually be transitioned across multiple PRs. Before any function is converted a collection of unit tests will be created so we can avoid regressing.
